File: golang/14-messenger/notification_client.go
package kata

type NotifyResult struct {
	Status     string
	Reason     string
	DeliveryID string
	Payload    map[string]any
	Audit      []string
}
type NotificationBackend struct{}

func (b NotificationBackend) send(p map[string]any) NotifyResult {
	if p["channel"] == "fax" {
		return NotifyResult{Status: "failed", Reason: "unsupported_channel", Payload: p}
	}
	return NotifyResult{Status: "sent", DeliveryID: p["channel"].(string) + "-" + p["recipient"].(string), Payload: p}
}

type NotificationGateway struct {
	client NotificationBackend
}

func (g NotificationGateway) dispatch(p map[string]any) NotifyResult {
	return g.client.send(p)
}

type NotificationAudit struct{}

func (a NotificationAudit) record(event string, p map[string]any) string {
	return event + ":" + p["channel"].(string)
}

type NotificationClient struct {
	gateway NotificationGateway
	audit   NotificationAudit
}

func NewNotificationClient() NotificationClient {
	return NotificationClient{NotificationGateway{NotificationBackend{}}, NotificationAudit{}}
}

func (c NotificationClient) send(p map[string]any) NotifyResult {
	message, ok := p["message"].(string)
	if !ok || message == "" {
		return NotifyResult{Status: "rejected", Reason: "missing_message", Payload: p}
	}

	normalized := map[string]any{
		"recipient": "unknown",
		"message":   message,
		"channel":   "email",
		"priority":  "normal",
	}
	for key, value := range p {
		normalized[key] = value
	}
	result := c.gateway.dispatch(normalized)
	result.Audit = []string{c.audit.record("queued", normalized), c.audit.record(result.Status, normalized)}
	if result.Status == "failed" && normalized["priority"] == "high" {
		result.Status = "retrying"
		result.Audit = append(result.Audit, c.audit.record("retry_scheduled", normalized))
	}
	return result
}

File: python/14-messenger/src/notification_client.py
class NotificationBackend:
    def send(self, payload):
        if payload["channel"] == "fax":
            return {"status": "failed", "reason": "unsupported_channel", "payload": payload}
        return {"status": "sent", "delivery_id": f"{payload['channel']}-{payload['recipient']}", "payload": payload}


class NotificationGateway:
    def __init__(self):
        self.client = NotificationBackend()

    def dispatch(self, payload):
        return self.client.send(payload)


class NotificationAudit:
    def record(self, event, payload):
        return f"{event}:{payload['channel']}"


class NotificationClient:
    def __init__(self):
        self.gateway = NotificationGateway()
        self.audit = NotificationAudit()

    def send(self, payload):
        if not payload.get("message"):
            return {"status": "rejected", "reason": "missing_message", "payload": payload}

        normalized = {
            "recipient": payload.get("recipient", "unknown"),
            "message": payload["message"],
            "channel": payload.get("channel", "email"),
            "priority": payload.get("priority", "normal"),
        }
        result = self.gateway.dispatch(normalized)
        result["audit"] = [self.audit.record("queued", normalized), self.audit.record(result["status"], normalized)]
        if result["status"] == "failed" and normalized["priority"] == "high":
            result["status"] = "retrying"
            result["audit"].append(self.audit.record("retry_scheduled", normalized))
        return result
